File: Word-Formed/src/com/google/corrigan/owen/wordformed/GameThread.java
package com.google.corrigan.owen.wordformed;

import android.graphics.Canvas;
import android.util.Log;
import android.view.SurfaceHolder;

//Thread for the Game to run on.
public class GameThread extends Thread
{
	//Constants to maintain a smooth rate of play
	private final static int MAX_FPS = 50;
	private final static int MAX_FRAME_SKIPS = 5;
	private final static int FRAME_PERIOD = 1000/MAX_FPS;
	private static final String TAG = GameThread.class.getSimpleName(); //Logcat
	private boolean running;
	private SurfaceHolder surfaceHolder; //Holders the SurfaceView
	private SinglePlayerGameView gamePanel;//Main Game
		
	public GameThread(SurfaceHolder surfaceHolder, SinglePlayerGameView gamePanel)
	{
		super();
		this.surfaceHolder = surfaceHolder;
		this.gamePanel = gamePanel;
	}
	public void setRunning(boolean running)
	{
		this.running = running;
	}
	
	@Override
	public synchronized void run()
	{
		//Boilerplate Initializations
		Canvas canvas;
		long beginTime;
		long timeDiff;
		int sleepTime;
		int framesSkipped;
		
		long ticks = 0L;
		while(running) //While Game runs
		{
			canvas = null;
			try {
				canvas = this.surfaceHolder.lockCanvas();//Lock the Canvas
				synchronized (surfaceHolder) {
					beginTime = System.currentTimeMillis();//Begin timing
					framesSkipped = 0;
					//this.gamePanel.update(); 
					this.gamePanel.render(canvas); //Render Canvas
					timeDiff = System.currentTimeMillis() - beginTime;//Time for rendering
					sleepTime = (int)(FRAME_PERIOD - timeDiff);
					
					if(sleepTime > 0)
					{
						try {
							Thread.sleep(sleepTime);
						} catch (InterruptedException e)
						{}
					}
					
					while(sleepTime < 0 && framesSkipped == MAX_FRAME_SKIPS)
					{
						this.gamePanel.update();
						sleepTime += FRAME_PERIOD;
						framesSkipped++;
					}
				}
			} finally {
				if(canvas != null) {
					surfaceHolder.unlockCanvasAndPost(canvas);
				}
			}
			ticks++;
			
			
		}
		Log.d(TAG, "Game loops "+ticks+" times.");
	}
}

File: app/src/main/java/com/google/corrigan/owen/wordformed/GameThread.kt
package com.google.corrigan.owen.wordformed

import android.graphics.Canvas
import android.util.Log
import android.view.SurfaceHolder
import kotlinx.coroutines.*

class GameThread(
    private val surfaceHolder: SurfaceHolder,
    private val gamePanel: SinglePlayerGameView
) {
    private val TAG = GameThread::class.java.simpleName
    private var running = false
    private var gameJob: Job? = null
    private val scope = CoroutineScope(Dispatchers.Main + SupervisorJob())

    companion object {
        private const val MAX_FPS = 50
        private const val FRAME_PERIOD = 1000 / MAX_FPS
    }

    fun setRunning(running: Boolean) {
        this.running = running
        if (running) {
            startLoop()
        } else {
            stopLoop()
        }
    }

    private fun startLoop() {
        if (gameJob?.isActive == true) return
        
        gameJob = scope.launch {
            var ticks = 0L
            while (running && !gamePanel.finished) {
                val beginTime = System.currentTimeMillis()
                
                var canvas: Canvas? = null
                try {
                    canvas = surfaceHolder.lockCanvas()
                    if (canvas != null) {
                        synchronized(surfaceHolder) {
                            gamePanel.render(canvas)
                        }
                    }
                } catch (e: Exception) {
                    Log.e(TAG, "Error in game loop", e)
                } finally {
                    canvas?.let {
                        surfaceHolder.unlockCanvasAndPost(it)
                    }
                }

                val timeDiff = System.currentTimeMillis() - beginTime
                val sleepTime = FRAME_PERIOD - timeDiff

                if (sleepTime > 0) {
                    delay(sleepTime)
                }
                
                ticks++
            }
            Log.d(TAG, "Game loops $ticks times.")
        }
    }

    private fun stopLoop() {
        running = false
        gameJob?.cancel()
    }
}
